components/scheduler/scheduler.go

```diff
@@ -0,0 +1,414 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package scheduler
+
+import (
+	"cmp"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	schedpb "go.temporal.io/api/schedule/v1"
+	workflowpb "go.temporal.io/api/workflow/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	schedspb "go.temporal.io/server/api/schedule/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/util"
+	"go.temporal.io/server/service/worker/scheduler"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type Scheduler struct {
+	*schedspb.HsmSchedulerState
+
+	tweakables     scheduler.TweakablePolicies
+	specBuilder    *scheduler.SpecBuilder
+	cspec          *scheduler.CompiledSpec
+	frontendClient workflowservice.WorkflowServiceClient
+	historyClient  resource.HistoryClient
+
+	logger  log.Logger
+	metrics metrics.Handler
+}
+
+func (s *Scheduler) resolveOverlapPolicy(overlapPolicy enumspb.ScheduleOverlapPolicy) enumspb.ScheduleOverlapPolicy {
+	if overlapPolicy == enumspb.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED {
+		overlapPolicy = s.Args.Schedule.Policies.OverlapPolicy
+	}
+	if overlapPolicy == enumspb.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED {
+		overlapPolicy = enumspb.SCHEDULE_OVERLAP_POLICY_SKIP
+	}
+	return overlapPolicy
+}
+
+func (s *Scheduler) populateTransientFieldsIfAbsent(logger log.Logger, handler metrics.Handler, specBuilder *scheduler.SpecBuilder, frontendClient workflowservice.WorkflowServiceClient, historyClient resource.HistoryClient) {
+	if s.logger == nil {
+		s.logger = logger
+	}
+	if s.metrics == nil {
+		s.metrics = handler
+	}
+	if s.frontendClient == nil {
+		s.frontendClient = frontendClient
+	}
+	if s.historyClient == nil {
+		s.historyClient = historyClient
+	}
```

Comment:
I don't think you want this on the HSM scheduler. This impure code should be in the executors.

The state machine should contain just state and transition definitions (pure logic).